Bring EXTRACT into alignment with PostgreSQL v14

jkosh44 opened the pull request:
Before this commit, EXTRACT was treated as an alias to date_part with slightly
different syntax. Both of these functions returned a float data type. PostgreSQL
v14 updated EXTRACT to return a numeric data type, which makes it compliant with
the SQL standard. This commit updates the EXTRACT function so that it return a
numeric data type so that it matches PostgreSQL. date_part still returns a float.

Additionally PostgreSQL v14 implemented EXTRACT explicitly for the DATE data
type. Previously DATEs were casted to TIMESTAMPs before extracting fields from
them. This commit also explicitly implements EXTRACT for DATE types, so they
aren't cast to a TIMESTAMP first. A consequence of this is that time related
fields (e.g. SECOND) are now rejected when trying to EXTRACT a DATE type.
However, The implementation for date_part still casts DATE types to TIMESTAMP
types. This means that date_part does not reject time related fields for DATE
types. This implementation matches PostgreSQL.

The postgres commit that implements these changes can be found here: 
https://github.com/postgres/postgres/commit/a2da77cdb4661826482ebf2ddba1f953bc74afe4

This commit also implements extracting EPOCHs from TIME types, which wasn't
previously implemented.

Fixes #9853, #9870,  #10027

<!--

Describe the contents of the PR briefly but completely.

If you write detailed commit messages, it is acceptable to copy/paste them
here, or write "see commit messages for details." If there is only one commit
in the PR, GitHub will have already added its commit message above.

-->

### Motivation
This PR adds a known-desirable feature: https://github.com/MaterializeInc/materialize/issues/9853 and https://github.com/MaterializeInc/materialize/issues/9870

<!--

Which of the following best describes the motivation behind this PR?

  * This PR fixes a recognized bug: [Link to issue.]

  * This PR adds a known-desirable feature: [Link to issue.]

  * This PR fixes a previously unreported bug.

    [Describe the bug in detail, as if you were filing a bug report.]

  * This PR adds a feature that has not yet been specified.

    [Write a brief specification for the feature, including justification
     for its inclusion in Materialize, as if you were writing the original
     feature specification.]

   * This PR refactors existing code.

    [Describe what was wrong with the existing code, if it is not obvious.]

-->

### Tips for reviewer

The previous implementation has the following structure. Each type (`Interval`, `TimestampLike`, etc) implemented methods to extract out the various parts from the type and return an `f64`.  For example `Interval` had `fn extract_quarter(&self) -> f64`. These methods can be further broken down into two categories:
1. Methods that involve floating point arithmetic. These are epoch, second, millisecond, microsecond.
2. Methods that really output a signed/unsigned integer which is cast to an `f64` before returning. These are all the other date/time parts.

The `date_part_X_inner` functions (ex: `date_part_timestamp_inner`) would simply call these functions and wrap the result in a `Result<Datum<'a>, EvalError>`.

This PR changes it to the following structure. Type 1 methods are converted into generic methods that let them do either floating point arithmetic or Numeric arithmetic. Type 2 methods are converted to returning signed/unsigned integers directly. The `date_part_X_inner` methods are also made generic so that they can be called with either `f64` or `Numeric`. When a `date_part_X_inner` method calls a type 1 method, is uses the supplied generic type. When a `date_part_X_inner` method calls a type 2 method, it casts the result to the supplied generic type.

The goal was to try and avoid duplicating every method twice for the different return types. Though after finishing it feels slightly over-engineered so I'm curious to hear if anyone has thoughts/suggestions on the approach.

Also I'm very open to stylistic comments since I'm new to both Rust and the code base.

<!--

Leave some tips for your reviewer, like:

    * The diff is much smaller if viewed with whitespace hidden.
    * [Some function/module/file] deserves extra attention.
    * [Some function/module/file] is pure code movement and only needs a skim.

Delete this section if no tips.

-->

### Checklist

- [X] This PR has adequate test coverage / QA involvement has been duly considered.
- [X] This PR adds a release note for any [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/user/content/release-notes.md#what-changes-require-a-release-note).

jkosh44 commented on doc/user/content/sql/functions/date_part.md:
```diff
@@ -0,0 +1,75 @@
+
+## Signatures
+
+TODO: this svg doesn't exist
```
I'm not really sure how to generate these images, so any help there is appreciated. An alternative would be to just get rid of this page and add a note to the bottom of the `EXTRACT` docs.

sploiselle replied:
https://github.com/MaterializeInc/materialize/tree/main/doc/user#railroad-diagrams-for-sql-grammar

jkosh44 replied:
Thanks, this should be fixed now

jkosh44 commented on src/repr/src/adt/numeric.rs:
```diff
@@ -616,3 +616,22 @@ pub fn extract_typ_mod(typ_mod: &[u64]) -> Result<Option<u8>, anyhow::Error> {
+
+/// Used to do potentially lossy value-to-value conversions while consuming the input value. Useful
+/// for interoperability between Numeric and f64.
+pub trait LossyFrom<T>: Sized {
```
I wasn't super happy with this approach. The issue is that you can't do `f64::from(i)` if `i` is type `i64` since the conversion is potentially lossy. Which means I can't use the `From<i64` trait on generic methods. So if anyone has a better approach please let me know. Also if there's a better place to put this please let me know.

jkosh44 replied:
Also if you have a better name, since it's technically only lossy when used with `f64` and not `Numeric`, let me know.

jkosh44 replied:
I think d155340 cleans this up slightly with the `DecimalLike` trait.

jkosh44 commented on src/repr/src/adt/numeric.rs:
```diff
@@ -616,3 +616,36 @@ pub fn extract_typ_mod(typ_mod: &[u64]) -> Result<Option<u8>, anyhow::Error> {
+{
+    /// Used to do value-to-value conversions while consuming the input value. Depending on the
+    /// implementation it may be potentially lossy.
+    fn lossy_from(i: i64) -> Self;
```
Apparently I just found out that there's a similar RFC for Rust: https://github.com/rust-lang/rfcs/pull/2484

benesch commented:
So sorry for the delay here, @jkosh44! Looks like this slipped through the cracks. I've pinged a few relevant folks.

jkosh44 commented:
> So sorry for the delay here, @jkosh44! Looks like this slipped through the cracks. I've pinged a few relevant folks.

No worries and no rush.

mjibson commented on doc/user/content/sql/functions/date-part.md:
```diff
@@ -0,0 +1,74 @@
+    parent: 'sql-functions'
+---
+
+`date_part` is modeled on the traditional Ingres equivalent to the SQL-standard 
```
Can this doc string be changed to summarize what the function does? Mentioning Ingres (I have no idea what that is myself), the sql spec, or postgres compat are not useful if a reader doesn't know what those are. If you think that's important information to discuss, it should be below the return values section. Other functions have a Details section for discussing this kind of stuff.

mjibson replied:
Followup: I see that this is exactly how postgres documents date_part and that is maybe useful. So I guess ignore my comment and this can stay like it is?

jkosh44 replied:
Yeah I tried to model these docs somewhat after the equivalent postgres docs found here: https://www.postgresql.org/docs/14/functions-datetime.html#FUNCTIONS-DATETIME-EXTRACT.

Ingres (https://dbdb.io/db/ingres) is an older database that Postgres is was modeled on (Postgres -> Post Ingres), and it seems like the only reason Postgres even has the `date_part` function was for compatibility with Ingres. Though after Postgres chaged `EXTRACT` to return numeric, they mention in the discussion and PR that `date_part` can be used as a faster less precise version of `EXTRACT`.

I can change the beginning to 
> `date_part` returns some time component from a time-based value, such as the year from a Timestamp. It is modeled on the traditional Ingres...

I'm also happy to remove the Ingres and SQL-standard mentions if we don't care about them.

Tagging @sploiselle since there was another comment about Ingres elsewhere.

sploiselle replied:
I think removing the mentions is good, unless you can point someone to a doc/reference that will enhance their understanding of the API.

benesch replied:
Agreed. I can’t imagine that in 2022 that comparison is useful to anyone.  I’ll bet $20 that that line in the PostgreSQL document dates back to pre 2000.

benesch replied:
@#$%@ I was off by one year: github.com/postgres/postgres/commit/e6767a32053963a968f419256e3629ace5e8df2f

jkosh44 replied:
So I removed the mentions to Ingres and the SQL standard. Let me know what you think. I'm honestly not great at technical writing so feel free to make suggestions to grammar and content.

mjibson commented on doc/user/content/sql/functions/extract.md:
```diff
@@ -8,13 +8,15 @@ menu:
 
 `EXTRACT` returns some time component from a time-based value, such as the year from a Timestamp.
 
+See [`date_part`](../date-part) for the traditional Ingres equivalent function.
```
It makes sense to point people from date_part to extract because extract is the new preferred way. Does it make sense to not even link people from extract to date_part, and instead remove this sentence?

jkosh44 replied:
I removed this sentence.

sploiselle commented on src/expr/src/scalar/func.rs:
```diff
@@ -1713,32 +1719,43 @@ impl TimestampLike for chrono::DateTime<chrono::Utc> {
 }
 
-fn date_part_interval<'a>(a: Datum<'a>, interval: Interval) -> Result<Datum<'a>, EvalError> {
+fn extract_interval(a: Datum, interval: Interval) -> Result<Datum, EvalError> {
```
This can be nominally reduced by hoisting up the generic param's literalization up a level

```diff
diff --git a/src/expr/src/scalar/func.rs b/src/expr/src/scalar/func.rs
index 331e1acdb..3df122a53 100644
--- a/src/expr/src/scalar/func.rs
+++ b/src/expr/src/scalar/func.rs
@@ -36,7 +36,7 @@ use repr::adt::array::ArrayDimension;
 use repr::adt::datetime::{DateTimeUnits, Timezone};
 use repr::adt::interval::Interval;
 use repr::adt::jsonb::JsonbRef;
-use repr::adt::numeric::{self, DecimalLike, Numeric};
+use repr::adt::numeric::{self, Dec, DecimalLike, Numeric};
 use repr::adt::regex::Regex;
 use repr::{strconv, ColumnName, ColumnType, Datum, DatumType, Row, RowArena, ScalarType};
 
@@ -1719,18 +1719,14 @@ impl TimestampLike for chrono::DateTime<chrono::Utc> {
     }
 }
 
-fn extract_interval(a: Datum, interval: Interval) -> Result<Datum, EvalError> {
-    let units = a.unwrap_str();
-    match units.parse() {
-        Ok(units) => date_part_interval_inner::<Numeric>(units, interval),
-        Err(_) => Err(EvalError::UnknownUnits(units.to_owned())),
-    }
-}
-
-fn date_part_interval(a: Datum, interval: Interval) -> Result<Datum, EvalError> {
+fn date_part_interval<'a, T>(a: Datum<'a>, b: Datum<'a>) -> Result<Datum<'a>, EvalError>
+where
+    T: DecimalLike + Into<Datum<'static>>,
+{
     let units = a.unwrap_str();
+    let interval = b.unwrap_interval();
     match units.parse() {
-        Ok(units) => date_part_interval_inner::<f64>(units, interval),
+        Ok(units) => date_part_interval_inner::<T>(units, interval),
         Err(_) => Err(EvalError::UnknownUnits(units.to_owned())),
     }
 }
@@ -2437,7 +2433,7 @@ impl BinaryFunc {
                 ))
             }
             BinaryFunc::ExtractInterval => {
-                eager!(|a, b: Datum| extract_interval(a, b.unwrap_interval()))
+                eager!(|a, b| date_part_interval::<Numeric>(a, b))
             }
             BinaryFunc::ExtractTime => {
                 eager!(|a, b: Datum| extract_time(a, b.unwrap_time()))
@@ -2452,7 +2448,7 @@ impl BinaryFunc {
                 eager!(|a, b: Datum| extract_date(a, b.unwrap_date()))
             }
             BinaryFunc::DatePartInterval => {
-                eager!(|a, b: Datum| date_part_interval(a, b.unwrap_interval()))
+                eager!(|a, b| date_part_interval::<f64>(a, b))
             }
             BinaryFunc::DatePartTime => {
                 eager!(|a, b: Datum| date_part_time(a, b.unwrap_time()))
```

I think you can do the same thing for time, and almost the same thing for timestamp/tz; the latter just requires the existing unwrap.

jkosh44 replied:
Just pushed this change.

sploiselle commented on src/expr/src/scalar/func.rs:
```diff
@@ -3626,10 +3752,25 @@ impl UnaryFunc {
-            DatePartTimestamp(units) => date_part_timestamp_inner(*units, a.unwrap_timestamp()),
-            DatePartTimestampTz(units) => date_part_timestamp_inner(*units, a.unwrap_timestamptz()),
+            ExtractInterval(units) => {
+                date_part_interval_inner::<Numeric>(*units, a.unwrap_interval())
```
This was not your doing, but all of these unwraps could just get moved inside this inner func:

```diff
diff --git a/src/expr/src/scalar/func.rs b/src/expr/src/scalar/func.rs
index 331e1acdb..683d941d6 100644
--- a/src/expr/src/scalar/func.rs
+++ b/src/expr/src/scalar/func.rs
@@ -1735,13 +1735,11 @@ fn date_part_interval(a: Datum, interval: Interval) -> Result<Datum, EvalError>
     }
 }
 
-fn date_part_interval_inner<T>(
-    units: DateTimeUnits,
-    interval: Interval,
-) -> Result<Datum<'static>, EvalError>
+fn date_part_interval_inner<T>(units: DateTimeUnits, a: Datum) -> Result<Datum<'static>, EvalError>
 where
     T: DecimalLike + Into<Datum<'static>>,
 {
+    let interval = a.unwrap_interval();
     match units {
         DateTimeUnits::Epoch => Ok(interval.as_seconds::<T>().into()),
         DateTimeUnits::Millennium => Ok(T::from(interval.millennia()).into()),
@@ -3752,9 +3750,7 @@ impl UnaryFunc {
             CharLength => char_length(a),
             IsRegexpMatch(regex) => Ok(is_regexp_match_static(a, &regex)),
             RegexpMatch(regex) => regexp_match_static(a, temp_storage, &regex),
-            ExtractInterval(units) => {
-                date_part_interval_inner::<Numeric>(*units, a.unwrap_interval())
-            }
+            ExtractInterval(units) => date_part_interval_inner::<Numeric>(*units, a),
             ExtractTime(units) => date_part_time_inner::<_, Numeric>(*units, a.unwrap_time()),
             ExtractTimestamp(units) => {
                 date_part_timestamp_inner::<_, Numeric>(*units, a.unwrap_timestamp())
@@ -3763,7 +3759,7 @@ impl UnaryFunc {
                 date_part_timestamp_inner::<_, Numeric>(*units, a.unwrap_timestamptz())
             }
             ExtractDate(units) => extract_date_inner(*units, a.unwrap_date()),
-            DatePartInterval(units) => date_part_interval_inner::<f64>(*units, a.unwrap_interval()),
+            DatePartInterval(units) => date_part_interval_inner::<f64>(*units, a),
             DatePartTime(units) => date_part_time_inner::<_, f64>(*units, a.unwrap_time()),
             DatePartTimestamp(units) => {
                 date_part_timestamp_inner::<_, f64>(*units, a.unwrap_timestamp())
```

note that I don't think this blithely works with some of the other suggestions

jkosh44 replied:
Just pushed this change for `interval`, `time`, and `date`.

sploiselle reviewed (commented):
I'm a soft LGTM on this right now, modulo the comments; I need to take another look at it this evening.

Feel free to hold off on implementing anything mentioned here until I (or someone else) comes back with firmer suggestions. I agree this feels a little over-engineered, but I think that's maybe just an artifact of the trait.

jkosh44 commented on doc/user/content/sql/functions/extract.md:
```diff
@@ -8,13 +8,15 @@ menu:
 
 `EXTRACT` returns some time component from a time-based value, such as the year from a Timestamp.
 
+See [`date_part`](../date_part) for the traditional Ingres equivalent function.
```
I've removed this sentence.

jkosh44 commented:
@sploiselle @mjibson Just pushed some changes that address your comments and fix merge conflicts. Thanks for the review!

mjibson commented on doc/user/content/sql/functions/date-part.md:
````diff
@@ -0,0 +1,74 @@
+
+```sql
+SELECT date_part('CENTURIES', DATE '2006-01-02')
+AS sec_extr;
````
This isn't seconds so is confusingly named. If it were me I'd take out the entire `AS` statement here and above.

jkosh44 replied:
Just removed them. I had originally just blindly copied this from the extract docs, so I removed if form the extract docs too.

mjibson reviewed (approved):
Docs changes LGTM. Other folks might have opinions about the Rust, I don't.

sploiselle commented:
@jkosh44 Sorry for the delay here; I saw there were some failing tests and forgot to circle back here. Overall this approach LGTM; if there's some simplification that becomes obvious, can always refactor. When I came back to this today, though, I had an easy time reasoning about the changes, so I think it's in a good state.

LMK if you'd like any help chasing down these failures.

jkosh44 commented:
@sploiselle Sorry about that, I didn't notice I had failed tests. I just pushed some changes that should hopefully fix them, but I think I need someone to starts the build for me (I'm not a member of the build-authorized team).

jkosh44 commented:
@sploiselle Double apologies, I definitely should have confirmed locally that I had fixed everything. I fixed the remaining errors and confirmed locally.

# Lint Docs:
```
$ ./bin/ci-builder run stable ci/test/lint-docs.sh 
Start building sites … 
hugo v0.87.0-B0C541E4+extended linux/amd64 BuildDate=2021-08-03T10:57:28Z VendorInfo=gohugoio
WARN 2022/01/31 20:53:23 "/home/joe/IdeaProjects/materialize/doc/user/content/demos/_index.md:1:1": duplicate menu entry with identifier "Try Materialize" in menu "main"
WARN 2022/01/31 20:53:23 "/home/joe/IdeaProjects/materialize/doc/user/content/sql/_index.md:1:1": duplicate menu entry with identifier "SQL Overview" in menu "main"

                   | EN   
-------------------+------
  Pages            | 199  
  Paginator pages  |   0  
  Non-page files   |   0  
  Static files     |  27  
  Processed images |   0  
  Aliases          |  63  
  Sitemaps         |   1  
  Cleaned          |   0  

Total in 489 ms
Skipping the checking of external links.
htmltest started at 08:53:23 on ci/www/public
========================================================================
✔✔✔ passed in 212.259593ms
tested 240 documents
```

# Testdrive

```
$ cargo run --bin testdrive --release -- test/testdrive/date_func.td 
    Finished release [optimized + debuginfo] target(s) in 0.19s
     Running `target/release/testdrive test/testdrive/date_func.td`
Configuration parameters:
    Kafka address: localhost:9092
    Schema registry URL: http://localhost:8081/
    materialized host: Tcp("localhost")
    error limit: 10
--- test/testdrive/date_func.td
Running test 1 time(s) ... 
Run 1 ...
> DROP DATABASE materialize
> SELECT EXTRACT(NULL FROM CAST('2011-11-11' AS DATE));
query error matches; continuing
> SELECT EXTRACT(NULL FROM CAST('11:11:11' AS TIME));
query error matches; continuing
> SELECT EXTRACT(NULL FROM CAST('2011-11-11' AS TIMESTAMP));
query error matches; continuing
> SELECT EXTRACT(NULL FROM CAST('2011-11-11' AS TIMESTAMP WITH TIME ZONE));
query error matches; continuing
> SELECT DATE_PART(NULL, CAST('2011-11-11' AS DATE)) IS NULL;
rows match; continuing at ts 1643663789.6457105
> SELECT DATE_PART(NULL, CAST('11:11:11' AS TIME)) IS NULL;
rows match; continuing at ts 1643663789.6461234
> SELECT DATE_PART(NULL, CAST('2011-11-11' AS TIMESTAMP)) IS NULL;
rows match; continuing at ts 1643663789.646539
> SELECT DATE_PART(NULL, CAST('2011-11-11' AS TIMESTAMP WITH TIME ZONE)) IS NULL;
rows match; continuing at ts 1643663789.6470332
> SELECT EXTRACT('foo' FROM CAST('2011-11-11' AS DATE));
query error matches; continuing
> SELECT EXTRACT('foo' FROM CAST('11:11:11' AS TIME));
query error matches; continuing
> SELECT EXTRACT('foo' FROM CAST('2011-11-11' AS TIMESTAMP));
query error matches; continuing
> SELECT EXTRACT('foo' FROM CAST('2011-11-11' AS TIMESTAMP WITH TIME ZONE));
query error matches; continuing
> SELECT DATE_PART('foo', CAST('2011-11-11' AS DATE));
query error matches; continuing
> SELECT DATE_PART('foo', CAST('11:11:11' AS TIME));
query error matches; continuing
> SELECT DATE_PART('foo', CAST('2011-11-11' AS TIMESTAMP));
query error matches; continuing
> SELECT DATE_PART('foo', CAST('2011-11-11' AS TIMESTAMP WITH TIME ZONE));
query error matches; continuing
> SELECT EXTRACT('' FROM CAST('2011-11-11' AS DATE));
query error matches; continuing
> SELECT EXTRACT('' FROM CAST('11:11:11' AS TIME));
query error matches; continuing
> SELECT EXTRACT('' FROM CAST('2011-11-11' AS TIMESTAMP));
query error matches; continuing
> SELECT EXTRACT('' FROM CAST('2011-11-11' AS TIMESTAMP WITH TIME ZONE));
query error matches; continuing
> SELECT DATE_PART('', CAST('2011-11-11' AS DATE));
query error matches; continuing
> SELECT DATE_PART('', CAST('11:11:11' AS TIME));
query error matches; continuing
> SELECT DATE_PART('', CAST('2011-11-11' AS TIMESTAMP));
query error matches; continuing
> SELECT DATE_PART('', CAST('2011-11-11' AS TIMESTAMP WITH TIME ZONE));
query error matches; continuing
> SELECT EXTRACT('second' FROM CAST('2011-11-11' AS DATE));
query error matches; continuing
> SELECT EXTRACT('second' FROM CAST('2011-11-11' AS TIMESTAMP));
rows match; continuing at ts 1643663789.6538699
> SELECT EXTRACT('second' FROM CAST('2011-11-11' AS TIMESTAMP WITH TIME ZONE));
rows match; continuing at ts 1643663789.6543183
> SELECT DATE_PART('second', CAST('2011-11-11' AS DATE));
rows match; continuing at ts 1643663789.6547601
> SELECT DATE_PART('second', CAST('2011-11-11' AS TIMESTAMP));
rows match; continuing at ts 1643663789.6551979
> SELECT DATE_PART('second', CAST('2011-11-11' AS TIMESTAMP WITH TIME ZONE));
rows match; continuing at ts 1643663789.6555967
> SELECT EXTRACT('day' FROM CAST(NULL AS DATE)) IS NULL;
rows match; continuing at ts 1643663789.6559987
> SELECT EXTRACT('day' FROM CAST(NULL AS TIME)) IS NULL;
rows match; continuing at ts 1643663789.6563857
> SELECT EXTRACT('day' FROM CAST(NULL AS TIMESTAMP)) IS NULL;
rows match; continuing at ts 1643663789.656778
> SELECT EXTRACT('day' FROM CAST(NULL AS TIMESTAMP WITH TIME ZONE)) IS NULL;
rows match; continuing at ts 1643663789.6571867
> SELECT DATE_PART('day', CAST(NULL AS DATE)) IS NULL;
rows match; continuing at ts 1643663789.6576157
> SELECT DATE_PART('day', CAST(NULL AS TIME)) IS NULL;
rows match; continuing at ts 1643663789.6580007
> SELECT DATE_PART('day', CAST(NULL AS TIMESTAMP)) IS NULL;
rows match; continuing at ts 1643663789.6583924
> SELECT DATE_PART('day', CAST(NULL AS TIMESTAMP WITH TIME ZONE)) IS NULL;
rows match; continuing at ts 1643663789.6588137
+++ testdrive completed successfully.
```

sploiselle commented:
No apologies necessary! Testing is iterative
